roachtest: optimize cluster destroy and create

nameisbhaskar opened the pull request:
Currently, roachtest creates a cluster for running tests. Once the test is complete, it tries to reuse the cluster for the next test. Incase case the cluster cannot be reused for the next test, it destroys and creates a new cluster. This happens in sequence today.
cluster destory takes around a minute or two and the same is with cluster create. So, overall, this operation takes away ~4 minutes of the roachtest time.
If we run destroy in a goroutine and start the cluster create for the next test, we should be able to save 2 minutes.

Considering a teamcity nightly run, we see that we create/destroy ~20 clusters in a worker, which accounts to 2*20 = 40 minutes!!

Fixes: #125228
Epic: None

nameisbhaskar committed:
roachtest: optimize cluster destroy and create

Currently, roachtest creates a cluster for running tests. Once the test is complete, it tries to reuse the cluster for the next test. Incase case the cluster cannot be reused for the next test, it destroys and creates a new cluster. This happens in sequence today.
cluster destory takes around a minute or two and the same is with cluster create. So, overall, this operation takes away ~4 minutes of the roachtest time.
If we run destroy in a goroutine and start the cluster create for the next test, we should be able to save 2 minutes.

Considering a teamcity nightly run, we see that we create/destroy ~20 clusters in a worker, which accounts to 2*20 = 40 minutes!!

Fixes: #125228
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125245)
<!-- Reviewable:end -->

herkolategan commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -605,17 +605,24 @@ func (r *testRunner) runWorker(
 			if ctx.Err() != nil {
 				// The context has been canceled. No need to continue.
+				r.destroyCluster(ctx, clusterDestroyWg, c, l, true)
+				clusterDestroyWg.Wait() // wait for the clusters to be destroyed
```
It seems like the `waitgroup wait` is added before all possible return points. Is there any reason for not moving it a `defer` above, and creating the waitgroup a little earlier so it can be part of that `defer`, so if someone adds another return point they wouldn't have to keep this in mind? Similar to how the `quota` is release.

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -926,6 +933,34 @@ func (r *testRunner) runWorker(
+	l *logger.Logger,
+	wait bool,
+) {
+	if c == nil {
```
Is there a case where the cluster would be nil?

nameisbhaskar replied:
This is a prcaution, in case the cluster is passed as nil. Currently, the code path should hit this condition.

nameisbhaskar replied:
There was one flow in which there was a possibility of cluster being nil because of a if condition, where I added this. The scenario is no more now, but, I kept it just in case.

srosenberg replied:
If it's _provable_ that `c` cannot be nil, then it's best to make it a _precondition_.

nameisbhaskar replied:
Sure!

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -605,17 +605,22 @@ func (r *testRunner) runWorker(
 		select {
 		case <-interrupt:
 			l.ErrorfCtx(ctx, "worker detected interruption")
+			r.destroyCluster(ctx, clusterDestroyWg, c, l, true)
```
nit: if you do keep the `wait` flag, then add a `/* Wait */` comment.

nameisbhaskar replied:
wait is removed as per your suggestion

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -926,6 +933,34 @@ func (r *testRunner) runWorker(
+		return
+	}
+	if c.IsLocal() || wait {
+		// local cluster - do not use goroutine as we generally recreate the cluster.
```
I don't quite follow this comment. Is it because you can only have one local cluster so it must be done sequentially?

nameisbhaskar replied:
Yes, without this condition CI fails as it uses local cluster.

srosenberg replied:
Let's state that explicitly in the comment; instead of the above, something like this,

> N.B. multiple local clusters aren't supported, hence we must use a blocking call.

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -926,6 +933,34 @@ func (r *testRunner) runWorker(
+	clusterDestroyWg *sync.WaitGroup,
+	c *clusterImpl,
+	l *logger.Logger,
+	wait bool,
```
Instead of passing a bool to cover this edge case, what if you just called `c.Destroy(context.Background(), closeLogger, l)` directly when you don't want to wait?

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -605,17 +605,22 @@ func (r *testRunner) runWorker(
 		}
 	}()
-
+	// cluster destroy can be done concurrently. We just need to make sure that we exit the roachtest once all
```
nit: exit the roachtest runner?

DarrylWong reviewed (commented):
Nice! I do wonder if we will run into any issues wrt quota if we're now trying to create resources before they are freed on the cloud end. Maybe the built in retry would be enough time to get past that though.

Should we also run a smoke test before merging?

nameisbhaskar commented:
> Nice! I do wonder if we will run into any issues wrt quota if we're now trying to create resources before they are freed on the cloud end. Maybe the built in retry would be enough time to get past that though.
> 
> Should we also run a smoke test before merging?

I ran a nightly build from this. I will run another one and paste the same here!

nameisbhaskar commented:
Thanks @herkolategan and @DarrylWong for reviewing this! I have addressed the comments. Please do take a look.

I will run one nightly build from this PR to see all runs ok.

nameisbhaskar commented:
https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15600985?hideProblemsFromDependencies=false&hideTestsFromDependencies=false

srosenberg commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -605,17 +605,22 @@ func (r *testRunner) runWorker(
 		select {
 		case <-interrupt:
 			l.ErrorfCtx(ctx, "worker detected interruption")
+			c.Destroy(context.Background(), closeLogger, l)
```
What's the rationale for adding a blocking call on these paths? Note that generally speaking this call has no upper bound on the running time.

nameisbhaskar replied:
makes sense. Will remove this and rely of GC to handle this.

srosenberg commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -605,17 +605,22 @@ func (r *testRunner) runWorker(
 		}
 	}()
-
+	// cluster destroy can be done concurrently. We just need to make sure that we exit the roachtest runner once all
```
Nit: comment can be made more precise. The `WaitGroup` just ensures that all pending Destroy calls have completed.

However, note that this may be introducing a new behavior; whereas previously an interrupted worker would exit abruptly, now it has to wait for Destroy(s) to finish.

nameisbhaskar replied:
This is correct. Till today, we were relying on GC to destroy the clusters in case of errors.

srosenberg commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -926,6 +933,30 @@ func (r *testRunner) runWorker(
+		defer clusterDestroyWg.Done()
+		// We use a context that can't be canceled for the Destroy().
+		ci.Destroy(context.Background(), closeLogger, l)
+		l.PrintfCtx(ctx, "cluster %s destroyed.", ci)
```
Nit: this seems redundant. We already log completion in `cluster.go` (`doDestroy`).

nameisbhaskar commented:
Thanks @srosenberg for the review.

srosenberg commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -716,6 +720,7 @@ func (r *testRunner) runWorker(
 		// case.
 		if err := VerifyLibraries(testToRun.spec.NativeLibs, arch); err != nil {
 			shout(ctx, l, stdout, "Library verification failed: %s", err)
+			c.Destroy(context.Background(), closeLogger, l)
```
I'm still fuzzy why we're adding `c.Destroy` to error paths? Don't we want to fail fast instead and let GC handle the clean-up at a later time?

srosenberg commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -605,7 +605,10 @@ func (r *testRunner) runWorker(
 		}
 	}()
-
+	// cluster destroy can be done concurrently.
```
Nit: a single line comment should suffice. What part of this code isn't self-documenting? Arguably, only the `defer` statement may need a supporting comment.

srosenberg commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -926,6 +927,24 @@ func (r *testRunner) runWorker(
+func (r *testRunner) destroyCluster(
+	clusterDestroyWg *sync.WaitGroup, c *clusterImpl, l *logger.Logger,
+) {
+	if c.IsLocal() {
```
I'd factor this case out and rename the above func to something like `destroyClusterAsync`. Otherwise, there is a confounding factor. The above comment (L666) says, "do not wait", yet we end up waiting in case of `c.IsLocal()`. Also, the `WaitGroup` is coupled with the case which doesn't need it, since it's a blocking call.

nameisbhaskar replied:
makes sense! I updated the comments. Thanks @srosenberg !

DarrylWong reviewed (approved):
LGTM

srosenberg reviewed (approved):
Let's do another, final smoke test run with SELECT_PROBABILITY=0.6, then merge.

nameisbhaskar commented:
Thanks @DarrylWong and @srosenberg !

I have started the nightly with SELCT_PROBABILITY=0.6.
https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15637536?hideProblemsFromDependencies=false&hideTestsFromDependencies=false

I will merge once this completes and all looks good!

nameisbhaskar commented:
bors r=@DarrylWong,@srosenberg
